include/secp256k1_musig.h

```diff
@@ -44,98 +47,54 @@ typedef struct {
- *
- * The workflow for this structure is as follows:
- *
- * 1. This structure is initialized with `musig_session_init` or
- *    `musig_session_init_verifier`, which initializes
- *    all other fields. The public session is initialized with the signers'
- *    nonce_commitments.
+/** Opaque data structure that holds a MuSig signature template.
  *
- * 2. In a non-public session the nonce_commitments are set with the function
- *    `musig_get_public_nonce`, which also returns the signer's public nonce. This
- *    ensures that the public nonce is not exposed until all commitments have been
- *    received.
- *
- * 3. Each individual data struct should be updated with `musig_set_nonce` once a
- *    nonce is available. This function takes a single signer data struct rather than
- *    an array because it may fail in the case that the provided nonce does not match
- *    the commitment. In this case, it is desirable to identify the exact party whose
- *    nonce was inconsistent.
- *
- * Fields:
- *   present: indicates whether the signer's nonce is set
- *     nonce: public nonce, must be a valid curvepoint if the signer is `present`
- * nonce_commitment: commitment to the nonce, or all-bits zero if a commitment
- *                   has not yet been set
+ *  Guaranteed to be 64 bytes in size. No serialization and parsing functions
+ *  (yet).
  */
 typedef struct {
-    int present;
-    secp256k1_xonly_pubkey nonce;
-    unsigned char nonce_commitment[32];
-} secp256k1_musig_session_signer_data;
+    unsigned char data[64];
+} secp256k1_musig_template;
 
-/** Opaque data structure that holds a MuSig partial signature.
+/** Opaque data structure that holds a cache for a MuSig session.
  *
- *  The exact representation of data inside is implementation defined and not
- *  guaranteed to be portable between different platforms or versions. It is however
- *  guaranteed to be 32 bytes in size, and can be safely copied/moved. If you need
- *  to convert to a format suitable for storage, transmission, or comparison, use the
- *  `musig_partial_signature_serialize` and `musig_partial_signature_parse`
- *  functions.
+ *  Guaranteed to be 65 bytes in size. No serialization and parsing functions
+ *  (yet).
  */
 typedef struct {
-    unsigned char data[32];
-} secp256k1_musig_partial_signature;
+    unsigned char data[65];
+} secp256k1_musig_session_cache;
+
+/** Guaranteed to be 64 bytes in size. This structure MUST NOT be copied or
+ *  read or written to it directly. A signer who is online throughout the whole
+ *  process and can keep this structure in memory can use the provided API
+ *  functions for a safe standard workflow. See
+ *  https://blockstream.com/2019/02/18/musig-a-new-multisignature-standard/ for
+ *  more details about the risks associated with serializing or deserializing
+ *  this structure. There are no serialization and parsing functions (yet).
+ */
+typedef struct {
+    unsigned char data[64];
+} secp256k1_musig_secnonce;
```

Comment:
Might be good to add a `secp256k1_musig_pubnonce` here, too. Would help with type-safety and understandability, similar to other arguments for strong types.

```c
typedef struct {
    unsigned char data[66];
} secp256k1_musig_pubnonce;
```

Reply:
Sorry, just getting back to this. `pubnonce` is different from other arguments that we add types for. The only thing that you do with the `pubnonce` is sending it to the other parties. Hence, we already serialize it. Otherwise, you'd have to call an extra serialization function that does nothing but a memcpy. For the same reason, we don't have a type for Schnorr signatures.